akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala

```diff
@@ -0,0 +1,461 @@
+package akka.stream.scaladsl
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.Done
+import akka.stream.ActorMaterializer
+import akka.stream.testkit.StreamSpec
+import akka.stream.testkit.scaladsl.{ TestSink, TestSource }
+import akka.testkit.DefaultTimeout
+
+import scala.concurrent.Promise
+import scala.concurrent.duration._
+import scala.util.{ Failure, Success }
+
+class RestartSpec extends StreamSpec with DefaultTimeout {
+
+  implicit val mat = ActorMaterializer()
+  import system.dispatcher
+
+  "A restart with backoff source" should {
+    "run normally" in {
+      val created = new AtomicInteger()
+      val probe = RestartSource.withBackoff(10.millis, 20.millis, 0) { () ⇒
+        created.incrementAndGet()
+        Source.repeat("a")
+      }.runWith(TestSink.probe)
+
+      probe.requestNext("a")
+      probe.requestNext("a")
+      probe.requestNext("a")
+      probe.requestNext("a")
+      probe.requestNext("a")
+
+      created.get() should ===(1)
+    }
+
+    "restart on completion" in {
+      val created = new AtomicInteger()
+      val probe = RestartSource.withBackoff(10.millis, 20.millis, 0) { () ⇒
+        created.incrementAndGet()
+        Source(List("a", "b"))
+      }.runWith(TestSink.probe)
+
+      probe.requestNext("a")
+      probe.requestNext("b")
+      probe.requestNext("a")
+      probe.requestNext("b")
+      probe.requestNext("a")
+
+      created.get() should ===(3)
```

Comment:
Isn't this a problematic shortcoming, that a stream using `RestartSource.withBackoff` can never complete based on the source completing? It is not what I would spontaneously expect from such a stage.